File: apps/bridge/src/pairing.ts
import { generateTransferKey, randomId, sealWithTransferKey } from "../../../packages/core/crypto";
import {
  installClaudeHook,
  installCodexHook,
  installCursorHook,
  installMonitorHelper,
  type InstallResult,
} from "./install";
import {
  createPairing,
  DEFAULT_RELAY_URL,
  loadConfig,
  machineConfigPath,
  normalizeRelayUrl,
  phonePairingPath,
  saveConfig,
} from "./config";
import type { PeerConfig } from "../../../packages/core/relay-client";
import type { PairingJoin } from "../../../packages/protocol/messages/pairing-join";
import { clearPhoneSeen } from "./presence";

export const DEFAULT_RELAY = DEFAULT_RELAY_URL;
export const PAIRING_CODE_TTL_MINUTES = 15;

function validConfig(config: PeerConfig, role: PeerConfig["role"]): boolean {
  return config.role === role
    && [config.relayUrl, config.room, config.senderId, config.myPublicKey,
      config.mySecretKey, config.peerPublicKey].every((value) => value.length > 0);
}

/** Return the shared machine identity only when both local halves are reciprocal. */
export function reusablePairing(replace = false): PeerConfig | null {
  if (replace) return null;
  try {
    const machine = loadConfig(machineConfigPath());
    const phone = loadConfig(phonePairingPath());
    return validConfig(machine, "machine")
      && validConfig(phone, "phone")
      && machine.room === phone.room
      && machine.relayUrl === phone.relayUrl
      && machine.myPublicKey === phone.peerPublicKey
      && machine.peerPublicKey === phone.myPublicKey
      ? machine
      : null;
  } catch {
    return null;
  }
}

export type OneTimePairing = {
  machineCfg: PeerConfig;
  phoneCfg: PeerConfig;
  mailboxId: string;
  transferKey: string;
  manualToken: string;
  httpBase: string;
  qrPayload: string;
  cursor: InstallResult | null;
  claude: InstallResult | null;
  codex: InstallResult | null;
  monitor: InstallResult | null;
};

export function relayHttpBase(relayUrl: string): string {
  const url = new URL(normalizeRelayUrl(relayUrl));
  url.protocol = url.protocol === "wss:" ? "https:" : "http:";
  return url.toString().replace(/\/$/, "");
}

/**
 * The QR contains a random mailbox id plus an independent 256-bit transfer key.
 * The mailbox id is the only part sent to the relay. The relay deletes the
 * ciphertext after the first successful GET and expires it after 15 minutes.
 */
export function oneTimePairingUri(relayUrl: string, mailboxId: string, transferKey: string): string {
  const query = new URLSearchParams({
    v: "2",
    u: relayHttpBase(relayUrl),
    m: mailboxId,
    k: transferKey,
  });
  return `granttap://pair-v2?${query.toString()}`;
}

/** Reciprocal machine + phone halves when both local files still match. */
export function reusablePairingHalves(): { machineCfg: PeerConfig; phoneCfg: PeerConfig } | null {
  const machineCfg = reusablePairing(false);
  if (!machineCfg) return null;
  try {
    const phoneCfg = loadConfig(phonePairingPath());
    return validConfig(phoneCfg, "phone") ? { machineCfg, phoneCfg } : null;
  } catch {
    return null;
  }
}

/** Same room on disk, even if the key halves drifted and are no longer reciprocal. */
export function parkedPairingHalves(): { machineCfg: PeerConfig; phoneCfg: PeerConfig } | null {
  const halves = reusablePairingHalves();
  if (halves) return halves;
  try {
    const machineCfg = loadConfig(machineConfigPath());
    const phoneCfg = loadConfig(phonePairingPath());
    return validConfig(machineCfg, "machine")
      && validConfig(phoneCfg, "phone")
      && machineCfg.room === phoneCfg.room
      && machineCfg.relayUrl === phoneCfg.relayUrl
      ? { machineCfg, phoneCfg }
      : null;
  } catch {
    return null;
  }
}

function machineIdentityPresent(): boolean {
  try {
    return validConfig(loadConfig(machineConfigPath()), "machine");
  } catch {
    return false;
  }
}

/**
 * A new computer mints a candidate room until a phone that is already in a
 * room scans its QR. That scan moves this computer into the phone's room —
 * it does not create a second room. Reconnect re-issues a QR for this room.
 * `replace` starts a different room. A Mesh invite is not this room.
 */

export type PairingJoinResult = "adopted" | "already" | "rejected";

/**
 * The phone already had a room. This computer keeps its machine keys and
 * switches into that room so the next heartbeat lands where the phone is.
 */
export function applyPairingJoin(payload: PairingJoin): PairingJoinResult {
  let machine: PeerConfig;
  try {
    machine = loadConfig(machineConfigPath());
  } catch {
    return "rejected";
  }
  if (!validConfig(machine, "machine")) return "rejected";
  if (payload.phoneCfg.room !== payload.room) return "rejected";
  if (payload.phoneCfg.myPublicKey !== payload.phonePublicKey) return "rejected";
  if (!validConfig(payload.phoneCfg, "phone")) return "rejected";
  let relayUrl: string;
  try {
    relayUrl = normalizeRelayUrl(payload.relayUrl);
  } catch {
    return "rejected";
  }
  if (machine.room === payload.room && machine.peerPublicKey === payload.phonePublicKey
      && machine.relayUrl === relayUrl) {
    return "already";
  }
  const nextMachine: PeerConfig = {
    ...machine,
    relayUrl,
    room: payload.room,
    peerPublicKey: payload.phonePublicKey,
    pushAuth: payload.phoneCfg.pushAuth ?? machine.pushAuth,
  };
  if (!validConfig(nextMachine, "machine")) return "rejected";
  saveConfig(machineConfigPath(), nextMachine);
  saveConfig(phonePairingPath(), payload.phoneCfg);
  clearPhoneSeen();
  return "adopted";
}
export async function createOneTimePairing(
  relayUrl: string,
  options: { installHooks?: boolean; replace?: boolean } = {},
): Promise<OneTimePairing> {
  const existing = options.replace ? null : parkedPairingHalves();
  if (!existing && !options.replace && machineIdentityPresent()) {
    throw new Error(
      "GrantTap already has a pairing room on this computer, but the local phone half is missing or does not match. Reconnect reuses that room. Confirm replace only to start a different room.",
    );
  }
  const { machineCfg, phoneCfg } = existing ?? createPairing(relayUrl);
  const mailboxId = randomId(16);
  const transferKey = generateTransferKey();
  const sealed = sealWithTransferKey(phoneCfg, transferKey);
  const httpBase = relayHttpBase(relayUrl);

  let response: Response;
  try {
    response = await fetch(`${httpBase}/pair/${mailboxId}`, {
      method: "PUT",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(sealed),
      signal: AbortSignal.timeout(10_000),
      redirect: "error",
    });
  } catch (error) {
    throw new Error(
      `The GrantTap relay is unavailable at ${httpBase}: ${error instanceof Error ? error.message : String(error)}`,
    );
  }
  if (!response.ok) {
    throw new Error(`The GrantTap relay rejected pairing with HTTP ${response.status}.`);
  }

  // Do not replace a working local pairing until the relay has accepted the
  // encrypted phone half. A failed onboarding attempt must be non-destructive.
  // Reconnect re-parks the same halves — writing them again would rotate backups
  // and look like a new room.
  if (!existing) {
    saveConfig(machineConfigPath(), machineCfg);
    saveConfig(phonePairingPath(), phoneCfg);
    clearPhoneSeen();
  }

  const installHooks = options.installHooks ?? process.env.GRANTTAP_SKIP_HOOKS !== "1";
  const cursor = installHooks ? installCursorHook() : null;
  const claude = installHooks ? installClaudeHook() : null;
  const codex = installHooks ? installCodexHook() : null;
  const monitor = installHooks ? installMonitorHelper() : null;

  return {
    machineCfg,
    phoneCfg,
    mailboxId,
    transferKey,
    manualToken: `${mailboxId}.${transferKey}`,
    httpBase,
    qrPayload: oneTimePairingUri(relayUrl, mailboxId, transferKey),
    cursor,
    claude,
    codex,
    monitor,
  };
}

File: crates/granttap-mcp/src/pairing.rs
//! One-time secure mailbox pairing (connect tool).

use crate::config::{
    create_pairing, machine_config_path, normalize_relay_url, phone_pairing_path, save_config,
    DEFAULT_RELAY, PAIRING_CODE_TTL_MINUTES,
};
use crate::crypto::{generate_transfer_key, random_id, seal_with_transfer_key};
use crate::http::put_json;
use blazingly_json::json;
use std::time::Duration;

pub struct OneTimePairing {
    pub http_base: String,
    pub qr_payload: String,
    pub manual_token: String,
    pub reused_existing: bool,
}

pub fn create_one_time_pairing(relay_url: Option<&str>) -> Result<OneTimePairing, String> {
    let env_relay = std::env::var("GRANTTAP_RELAY_URL")
        .or_else(|_| std::env::var("NODVOX_RELAY_URL"))
        .ok();
    let relay = normalize_relay_url(
        relay_url
            .or(env_relay.as_deref())
            .unwrap_or(DEFAULT_RELAY),
    )?;

    // Re-park existing pairing without rotating keys when machine.json already works.
    if let Ok(existing) = crate::config::load_config(&machine_config_path()) {
        if existing.relay_url == relay {
            if let Ok(phone) = crate::config::load_config(&phone_pairing_path()) {
                return park_phone_half(&relay, &phone, true);
            }
        }
    }

    let (machine_cfg, phone_cfg) = create_pairing(&relay)?;
    let parked = park_phone_half(&relay, &phone_cfg, false)?;
    save_config(&machine_config_path(), &machine_cfg)?;
    save_config(&phone_pairing_path(), &phone_cfg)?;
    let _ = crate::setup::install_monitor_helper();
    Ok(parked)
}

fn park_phone_half(
    relay: &str,
    phone_cfg: &crate::config::PeerConfig,
    reused_existing: bool,
) -> Result<OneTimePairing, String> {
    let mailbox_id = random_id(16)?;
    let transfer_key = generate_transfer_key()?;
    let phone_val = blazingly_json::to_value(phone_cfg).map_err(|e| e.to_string())?;
    let (nonce, boxed) = seal_with_transfer_key(&phone_val, &transfer_key)?;
    let body = blazingly_json::to_string(&json!({ "nonce": nonce, "box": boxed }))
        .map_err(|e| e.to_string())?;
    let http_base = relay_http_base(relay);
    let status = put_json(
        &format!("{http_base}/pair/{mailbox_id}"),
        &body,
        Duration::from_secs(10),
    )?;
    if !(200..300).contains(&status) {
        return Err(format!(
            "The GrantTap relay rejected pairing with HTTP {status}."
        ));
    }
    let qr_payload = one_time_pairing_uri(relay, &mailbox_id, &transfer_key);
    Ok(OneTimePairing {
        http_base,
        qr_payload,
        manual_token: format!("{mailbox_id}.{transfer_key}"),
        reused_existing,
    })
}

pub fn relay_http_base(relay_url: &str) -> String {
    let mut base = relay_url.to_string();
    if let Some(rest) = base.strip_prefix("wss://") {
        base = format!("https://{rest}");
    } else if let Some(rest) = base.strip_prefix("ws://") {
        base = format!("http://{rest}");
    }
    base.trim_end_matches('/').to_string()
}

pub fn one_time_pairing_uri(relay_url: &str, mailbox_id: &str, transfer_key: &str) -> String {
    format!(
        "granttap://pair-v2?v=2&u={}&m={mailbox_id}&k={transfer_key}",
        urlencoding_minimal(&relay_http_base(relay_url)),
    )
}

fn urlencoding_minimal(s: &str) -> String {
    let mut out = String::new();
    for b in s.bytes() {
        match b {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(b as char);
            }
            _ => out.push_str(&format!("%{b:02X}")),
        }
    }
    out
}

pub fn connect_text(pairing: &OneTimePairing) -> String {
    let header = if pairing.reused_existing {
        "Re-parked your existing GrantTap pairing (keys were not rotated). Scan or paste below on iPhone."
    } else {
        "Scan this QR with GrantTap on iPhone to pair this computer."
    };
    format!(
        "{header}\nRelay: {}\nSecure mailbox URI (scan QR, or paste in Pair GrantTap):\n{}\nManual secure token (mailbox.key): {}\nThe encrypted mailbox is single-use and expires after {} minutes.\nThe relay receives only a random mailbox id and ciphertext; the transfer key stays in this user-only QR/token.\n(Note: Rust MCP returns the URI as text; PNG QR is not generated yet.)",
        pairing.http_base,
        pairing.qr_payload,
        pairing.manual_token,
        PAIRING_CODE_TTL_MINUTES
    )
}
